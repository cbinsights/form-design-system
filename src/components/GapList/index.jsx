import React from 'react';
import cx from 'classnames';
import PropTypes from 'prop-types';

const spacingScale = (num) => {
  switch (num) {
    case 1:
      return 4;
    case 2:
      return 8;
    case 3:
      return 16;
    case 4:
      return 24;
    case 5:
      return 32;
    default:
      return 16;
  }
};

const GapList = ({ items, xGap = 2, yGap = 2, valign = 'top' }) => (
  <div style={{ margin: `-${spacingScale(yGap)}px -${spacingScale(xGap)}px` }}>
    {items.map((item, i) => (
      <div
        key={i}
        className={cx(
          'display--inlineBlock',
          `valign--${valign}`,
          `margin--x--${xGap}`,
          `margin--y--${yGap}`
        )}
      >
        {item}
      </div>
    ))}
  </div>
);

const steps = ['1', '2', '3', '4', '5'];

GapList.propTypes = {
  items: PropTypes.arrayOf(PropTypes.oneOfType([PropTypes.element, PropTypes.string]))
    .isRequired,
  xGap: PropTypes.oneOf(steps),
  /** Applies vertical margin given the FDS spacing scale */
  yGap: PropTypes.oneOf(steps),
  /** Nodes to render */
  /** Controls vertical-align property */
  valign: PropTypes.string,
  /** Applies horiztonal margin given the FDS spacing scale */
};

export default GapList;
