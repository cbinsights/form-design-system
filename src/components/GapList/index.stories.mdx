import { Meta, Story, Preview, Props } from "@storybook/addon-docs/blocks";
import {
  withKnobs,
  text,
  boolean,
  optionsKnob as options
} from "@storybook/addon-knobs";
import { arrayToOptions, ImportPath } from 'components/util/storybook';

import GapList from '.';
import Chip from 'components/Chip'

<Meta title="Components/GapList" decorators={[withKnobs]} component={GapList} />

# GapList
<ImportPath component={GapList} />

<Story name="Knobs" parameters={{ docs: { disable: true }}}>
  <div className="padding--all bgColor--aqua rounded--all">
    <GapList
      yGap={options("yPadding", arrayToOptions([1, 2, 3, 4, 5]), undefined, {
        display: "inline-radio"
      })}
      xGap={options("xPadding", arrayToOptions([1, 2, 3, 4, 5]), undefined, {
        display: "inline-radio"
      })}
      valign={options("valign", arrayToOptions(['top', 'center', 'bottom', 'baseline']), undefined, {
        display: "inline-radio"
      })}
      items={[
        <div class="media--l bgColor--white rounded--all elevation--1" />,
        <div class="media--xl bgColor--white rounded--all elevation--1" />,
        <div class="media--l bgColor--white rounded--all elevation--1" />,
        <div class="media--xl bgColor--white rounded--all elevation--1" />,
        <div class="media--l bgColor--white rounded--all elevation--1" />,
        <div class="media--xl bgColor--white rounded--all elevation--1" />,
        <div class="media--l bgColor--white rounded--all elevation--1" />,
        <div class="media--xl bgColor--white rounded--all elevation--1" />,
        <div class="media--xs bgColor--white rounded--all elevation--1" />,
        <div class="media--xl bgColor--white rounded--all elevation--1" />,
        <div class="media--l bgColor--white rounded--all elevation--1" />,
        <div class="media--xl bgColor--white rounded--all elevation--1" />,
        <div class="media--l bgColor--white rounded--all elevation--1" />,
        <div class="media--s bgColor--white rounded--all elevation--1" />,
        <div class="media--l bgColor--white rounded--all elevation--1" />,
      ]}
    />
  </div>
</Story>

<Preview isExpanded>
  <div className="padding--all bgColor--aqua rounded--all">
    <GapList
      items={[
        <div class="media--l bgColor--white rounded--all elevation--1" />,
        <div class="media--xl bgColor--white rounded--all elevation--1" />,
        <div class="media--l bgColor--white rounded--all elevation--1" />,
        <div class="media--xl bgColor--white rounded--all elevation--1" />,
        <div class="media--l bgColor--white rounded--all elevation--1" />,
        <div class="media--xl bgColor--white rounded--all elevation--1" />,
        <div class="media--l bgColor--white rounded--all elevation--1" />,
        <div class="media--xl bgColor--white rounded--all elevation--1" />,
        <div class="media--xs bgColor--white rounded--all elevation--1" />,
        <div class="media--xl bgColor--white rounded--all elevation--1" />,
        <div class="media--l bgColor--white rounded--all elevation--1" />,
        <div class="media--xl bgColor--white rounded--all elevation--1" />,
        <div class="media--l bgColor--white rounded--all elevation--1" />,
        <div class="media--s bgColor--white rounded--all elevation--1" />,
        <div class="media--l bgColor--white rounded--all elevation--1" />,
      ]}
    />
  </div>
</Preview>

Accepts an array of nodes and applies horizontal / vertical margin for you. 

- Use this when you want to space out elements and also want them to wrap!
- This only applies margin "between" nodes, so you must control any additional desired outside margin.
- all items get wrapped with an inline-block element.

<Props of={GapList} />

### Spacing text

Works just as well with text

<Preview>
  <Story name="Spacing text">
    <GapList
      items={[
        <p>lorem</p>,
        'ipsum',
        <b>dolor</b>,
        <small>sit</small>,
        <a href="http://cbinsights.com">amet</a>,
      ]}
    />
  </Story>
</Preview>

### Misc

<Preview>
  <Story name="Misc">
    <GapList
      items={[	
        <Chip key="horses" onClose={() => {}} label="Horses" theme="outline" />,	
        <Chip key="wombats" onClose={() => {}} label="Wombats" theme="outline" />,	
        <Chip key="seals" onClose={() => {}} label="Seals" theme="outline" />,	
        <Chip key="dogs" onClose={() => {}} label="Dogs" theme="outline" />,	
      ]}	
    />
  </Story>
</Preview>

