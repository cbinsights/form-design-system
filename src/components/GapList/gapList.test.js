import React from 'react';
import { shallow } from 'enzyme';

import GapList from '.';

describe('GapList component', () => {

  it('matches snapshot for string items', () => {
    const component = shallow(
      <GapList
        items={['why', 'did', 'it', 'have', 'to', 'be', 'snakes?']}
      />
    );
    expect(component).toMatchSnapshot();
  });

  it('matches snapshot for mixed strings and components', () => {
    const component = shallow(
      <GapList
        items={[
          <a key="ðŸ‘»" href="#">Spooky</a>,
          'snapshot',
        ]}
      />
    );
    expect(component).toMatchSnapshot();
  });

});
