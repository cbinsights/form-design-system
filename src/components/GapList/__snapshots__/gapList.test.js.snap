// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`GapList component matches snapshot for mixed strings and components 1`] = `
<div
  style={
    Object {
      "margin": "-8px -8px",
    }
  }
>
  <div
    className="display--inlineBlock valign--top margin--x--2 margin--y--2"
    key="0"
  >
    <a
      href="#"
      key="ðŸ‘»"
    >
      Spooky
    </a>
  </div>
  <div
    className="display--inlineBlock valign--top margin--x--2 margin--y--2"
    key="1"
  >
    snapshot
  </div>
</div>
`;

exports[`GapList component matches snapshot for string items 1`] = `
<div
  style={
    Object {
      "margin": "-8px -8px",
    }
  }
>
  <div
    className="display--inlineBlock valign--top margin--x--2 margin--y--2"
    key="0"
  >
    why
  </div>
  <div
    className="display--inlineBlock valign--top margin--x--2 margin--y--2"
    key="1"
  >
    did
  </div>
  <div
    className="display--inlineBlock valign--top margin--x--2 margin--y--2"
    key="2"
  >
    it
  </div>
  <div
    className="display--inlineBlock valign--top margin--x--2 margin--y--2"
    key="3"
  >
    have
  </div>
  <div
    className="display--inlineBlock valign--top margin--x--2 margin--y--2"
    key="4"
  >
    to
  </div>
  <div
    className="display--inlineBlock valign--top margin--x--2 margin--y--2"
    key="5"
  >
    be
  </div>
  <div
    className="display--inlineBlock valign--top margin--x--2 margin--y--2"
    key="6"
  >
    snakes?
  </div>
</div>
`;
